### PR TITLE
Show unlocked exercises for an exercise based on DB data

### PR DESCRIPTION
**Pull request by kntsoriano**
Closes https://github.com/exercism/exercism/issues/4223.

## Description
We originally show the unlocked exercises based on the exercise tree, but not on the currently unlocked solutions in the DB. If we change our unlocking logic, we'd need to change them in two places.

### CODE REVIEW

File: app/models/solution.rb [hunk 1]
```diff
@@ -71,7 +71,7 @@ def legacy?
   end
 
   def approved?
-    !!approved_by
+    auto_approve? || !!approved_by
```
[the remark this answers is not in the document]
Comment (kntsoriano) [reply app/models/solution.rb]: Good catch! We don't need this then :)

**Comment by kntsoriano** [comment]
@iHiD Can we merge?

File: test/models/exercise_with_unlocked_exercises_test.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  test "returns unlocked exercises for an exercise" do
+    exercise = create(:exercise)
+    unlocked_exercise = create(:exercise, unlocked_by: exercise)
+    locked_exercise = create(:exercise, unlocked_by: exercise)
+    user = create(:user)
+    create(:solution, exercise: unlocked_exercise, user: user)
+
+    exercise = ExerciseWithUnlockedExercises.new(exercise, user)
+
+    assert_delegated_equal [unlocked_exercise], exercise.unlocked_exercises
```
Comment (iHiD) [inline test/models/exercise_with_unlocked_exercises_test.rb]: Why is this not just `assert_equal`?

File: app/views/my/tracks/_core_exercises.html.haml [hunk 3]
```diff
@@ -21,13 +21,11 @@
                 .topic= topic_name
           .blurb= exercise.blurb
 
-        -if @side_exercises_and_solutions_by_unlocked_by[exercise.id].present? || (exercise.auto_approve? && @side_exercises_and_solutions_by_unlocked_by[nil].present?)
+        - if exercise.unlocked_exercises.any?
           .unlocked-exercises-section
             %h3 You've unlocked extra exercises!
             .unlocked-exercises
-              -if exercise.auto_approve?
-                =render "unlocked_exercises", unlocked_exercises_and_solutions: @side_exercises_and_solutions_by_unlocked_by[nil][0,10]
```
Comment (iHiD) [inline app/views/my/tracks/_core_exercises.html.haml]: Have we maintained this line somewhere? (ie if it's auto-approved we *also* show the first 10 side exercises with `unlocked_by: nil`)

File: app/models/exercise_map.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  end
+
+  def exercises
+    @exercises ||= track.
+      exercises.
+      map { |e| ExerciseWithSolution.new(e, solution_for(e)) }
+  end
+
+  def exercises_to_unlock_for(exercise)
+    exercise.unlocks.map do |exercise_to_unlock|
```
Comment (iHiD) [inline app/models/exercise_map.rb]: I think this is where we also need a line that says "if this exercise auto approves then all the side exercises with unlocked_by: false" go here.

**Comment by iHiD** [comment]
@kntsoriano I've left a couple of extra comments. Sorry for the delay. I have 25 PRs I'm working through reviewing today!

**Comment by sshine** [comment]
exercism/haskell#721 experienced something that sounds related to this after changing the way side-track exercises are unlocked in https://github.com/exercism/haskell/commit/596e99fecc95628d8022243d771771bcd41e4417

Could someone perhaps recommend a course of action if this is related?

**Comment by iHiD** [comment]
@sshine - Thanks. I need to run a script to recalculate everyone's tracks for Haskell. It's conceptually related to this PR, but it's a different codepath. I'll set that running for you now. It takes a few hours.

File: app/models/exercise_map.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    @exercises ||= track.
+      exercises.
+      map { |e| ExerciseWithSolution.new(e, solution_for(e)) }
+  end
+
+  def exercises_to_unlock_for(exercise)
+    if exercise.auto_approve?
+      return exercises.
+        select(&:side?).
+        select { |e| e.unlocked_by.blank? }[0, 10]
```
Comment (iHiD) [inline app/models/exercise_map.rb]: @kntsoriano I think this needs to be this **as well** as the bit below. An auto-approve might unlock exercises as well as the bonus ones.

**Comment by iHiD** [comment]
@kntsoriano Sorry this is taking a lot of iterations. I have headspace for this this week so hopefully we can get it merged down now. I've made one more comment.

**Pull request merged**